#79612 fix(client-go): return error in fake discovery

MaxHorstmann opened the pull request:
<!--  Thanks for sending a pull request!  Here are some tips for you:

1. If this is your first time, please read our contributor guidelines: https://git.k8s.io/community/contributors/guide/first-contribution.md#your-first-contribution and developer guide https://git.k8s.io/community/contributors/devel/development.md#development-guide
2. Please label this pull request according to what type of issue you are addressing, especially if this is a release targeted pull request. For reference on required PR/issue labels, read here:
https://git.k8s.io/community/contributors/devel/sig-release/release.md#issuepr-kind-label
3. Ensure you have added or ran the appropriate tests for your PR: https://git.k8s.io/community/contributors/devel/sig-testing/testing.md
4. If you want *faster* PR reviews, read how: https://git.k8s.io/community/contributors/guide/pull-requests.md#best-practices-for-faster-reviews
5. If the PR is unfinished, see how to mark it: https://git.k8s.io/community/contributors/guide/pull-requests.md#marking-unfinished-pull-requests
-->

#### What type of PR is this?

/kind bug

#### What this PR does / why we need it:

Fake discovery should return an error if an error-returning reactor was prepended. This is relevant e.g. for unit tests which test a function which relies on discovery to check if an API Server is available.

@qianlei90's created an earlier PR #86770 for this which didn't get merged. This is an up-to-date version of that PR.

#### Which issue(s) this PR fixes:

Fixes #79612 

#### Special notes for your reviewer:

Simple fix + unit test. 

Other fake types in client-go already have this behavior ([example](https://github.com/kubernetes/kubernetes/blob/master/staging/src/k8s.io/client-go/kubernetes/typed/core/v1/fake/fake_namespace.go#L60)). It's likely missing from fake Discovery due to an oversight.

#### Does this PR introduce a user-facing change?
NONE


#### Additional documentation e.g., KEPs (Kubernetes Enhancement Proposals), usage docs, etc.:
NONE

MaxHorstmann commented:
Apologies if this created any unwanted noise. This is a *draft* PR, work still in progress.

MaxHorstmann commented:
/kind bug

MaxHorstmann commented:
/retst

MaxHorstmann commented:
/retest

MaxHorstmann commented:
/release-note-none

alexzielenski commented:
/cc @Jefftree  can you take a look ty
/triage accepted

enj commented on staging/src/k8s.io/client-go/discovery/fake/discovery_test.go:
```diff
@@ -17,11 +17,14 @@ limitations under the License.
 
 import (
+	"errors"
+	"k8s.io/apimachinery/pkg/runtime"
```
Move the k8s import down into the block of hte other k8s imports.

MaxHorstmann replied:
ok, [done](https://github.com/kubernetes/kubernetes/pull/114291/commits/67fe0ed57b6d8d461dd8b9bf8eb3a0b719c21a62)

enj commented on staging/src/k8s.io/client-go/discovery/fake/discovery_test.go:
```diff
@@ -44,3 +47,19 @@ func TestFakingServerVersion(t *testing.T) {
+	if err == nil {
+		t.Fatal("ServerVersion should return error, returned nil instead")
+	}
+	if err.Error() != expectedError.Error() {
```
Direct comparison should work here (and if it does not, `errors.Is` is better than string comparison):

```suggestion
	if err != expectedError {
```

MaxHorstmann replied:
Confirmed, direct comparison does work.

MaxHorstmann commented:
@enj thanks! I addressed both your comments.

MaxHorstmann committed:
client-go: properly return error in fake discovery (#79612)

Fake discovery should return an error if an error-returning reactor was prepended.

This is relevant e.g. for unit tests which test a function which relies
on discovery to check if an API Server is available.

Matches existing behavior of other fake types, for example fake_namespace.

Adds unit test coverage as well.

enj commented:
/lgtm

enj commented:
/assign @sttts 

Easy review whenever you have the chance.

enj commented:
/milestone v1.27

sttts commented:
/approve

MaxHorstmann commented:
/retest-required
